src/ESLint.Formatter/sarif-with-rules.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @fileoverview SARIF v2.0 formatter
+ * @author Chris Meyer
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Helper Functions
+//------------------------------------------------------------------------------
+
+/**
+ * Returns the severity of warning or error
+ * @param {Object} message message object to examine
+ * @returns {string} severity level
+ * @private
+ */
+function getResultLevel(message) {
+    if (message.fatal || message.severity === 2) {
+        return "error";
+    }
+    return "warning";
+}
+
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+module.exports = function(results, rules) {
+    const sarifLog = {
+        version: "2.0.0",
+        $schema: "http://json.schemastore.org/sarif-2.0.0",
+        runs: [
+            {
+                tool: {
+                    name: "ESLint"
+                }
+            }
+        ]
+    };
+
+    const sarifFiles = {};
+    const sarifRules = {};
+    const sarifResults = [];
+
+    results.forEach(result => {
+
+        if (typeof sarifFiles[result.filePath] === "undefined") {
+
+            // Create a new entry in the files dictionary.
+            sarifFiles[result.filePath] = {
+                fileLocation: {
+                    uri: result.filePath
+                }
+            };
+        }
+
+        const messages = result.messages;
+
+        messages.forEach(message => {
+
+            const sarifResult = {
+                level: getResultLevel(message),
+                message: {
+                    text: message.message
+                },
+                locations: [
+                    {
+                        physicalLocation: {
+                            fileLocation: {
+                                uri: result.filePath
+                            }
+                        }
+                    }
+                ]
+            };
+
+            if (message.ruleId) {
+                sarifResult.ruleId = message.ruleId;
+
+                if (typeof sarifRules[message.ruleId] === "undefined") {
+                    const rule = rules.get(message.ruleId);
+
+                    // An unknown ruleId will return null. This check prevents unit test failure.
+                    if (rule) {
+
+                        // Create a new entry in the rules dictionary.
+                        sarifRules[message.ruleId] = {
```

Comment:
>sarifRules [](start = 24, length = 10)

rules dictionary will be an array soon

Reply:
Understood.

---
In reply to: [249139633](https://github.com/Microsoft/sarif-sdk/pull/1222#discussion_r249139633) [](ancestors = 249139633)